features.go

```diff
@@ -0,0 +1,185 @@
+// to allows future incompatible changes, or backward compatible changes.
+type featureFlag uint8
+
+// featureName represent the name of the feature and needed in order to have the
+// compile errors if we specify wrong feature name.
+type featureName string
+
+type nameToIndex map[featureName]int
+type indexToFlag map[int]featureFlag
+
+const (
+	// optionalFlag represent the feature which we might be used already but
+	// it isn't required yet, and if remote peer doesn't have this feature
+	// we may turn it off without disconnecting with peer.
+	optionalFlag featureFlag = 2 // 0b10
+
+	// requiredFlag represent the features which is required for proper
+	// peer interaction, we disconnect with peer if it doesn't have this
+	// particular feature.
+	requiredFlag featureFlag = 3 // 0b11
+)
+
+// featureVector represents the global/local feature vector. With this structure
+// you may set/get the feature by name and compare feature vector with remote
+// one.
+type featuresVector struct {
+	indexes  nameToIndex
+	features indexToFlag
+	compared uint32
+}
+
+// newFeatureVector creates new feature vector.
+func newFeatureVector(indexes nameToIndex) *featuresVector {
+	return &featuresVector{
+		indexes:  indexes,
+		features: make(indexToFlag),
+	}
+}
+
+// SetFeatureFlag assign flag to the feature.
+func (vector *featuresVector) SetFeatureFlag(name featureName,
+	flag featureFlag) error {
+
+	position, ok := vector.indexes[name]
+	if !ok {
+		return errors.Errorf("can't find feature with name: %v", name)
+	}
+
+	vector.features[position] = flag
+	return nil
+}
+
+// IsActive checks - is feature active or not, it might be disable during
+// comparision with remote feature vector if it was optional before and
+// remote peer doesn't support it.
+func (vector *featuresVector) IsActive(name featureName) (bool, error) {
+	if atomic.LoadUint32(&vector.compared) == 0 {
+		return false, errors.New("vector should be compared first.")
+	}
+
+	position, ok := vector.indexes[name]
+	if !ok {
+		return false, errors.Errorf("can't find feature with name: %v",
+			name)
+	}
+
+	_, exist := vector.features[position]
+	return exist, nil
+}
+
+// ToBytes turns feature vector into bytes array.
+func (vector *featuresVector) ToBytes() ([]byte, error) {
+	return encodeFeatures(vector)
+}
+
+// CheckCompatibility gets decoded feature vector as input, decode it and checks
+// that features are compatible. So if remote/local node doesn't have the feature
+// and local/remote node require it than vectors are incompatible.
+func (local *featuresVector) CheckCompatibility(data []byte) error {
+	remoteFlags := decodeFeatures(data)
+
+	for index, flag := range local.features {
+		if _, exist := remoteFlags[index]; !exist {
+			switch flag {
+			case requiredFlag:
+				return errors.New("Remote node haven't " +
+					"locally required feature")
+			case optionalFlag:
+				delete(local.features, index)
```

Comment:
Wow! You are right... Mutation of global feature vector might have led to very nasty bugs, also I like the idea of `SharedFeatures`.